File: Algorithms/43. Multiply Strings.java
class Solution {
    public String multiply(String num1, String num2) {
        if(num1.length()==0 || num2.length()==0) return "0";
        if(num1.length()<num2.length()){
            String temp=num1;
            num1=num2;
            num2=temp;
        }
        int[] ans=new int[1000];
        for(int i=num2.length()-1;i>=0;i--){
            int x=num2.charAt(i)-'0',carry=0;
            for(int j=num1.length()-1;j>=0;j--){
                int y=num1.charAt(j)-'0';
                int z=x*y+carry;
                carry = z/10;
                z=z%10;
                int index=num2.length()-1-i+num1.length()-1-j;
                ans[index]+=z;
                carry+=ans[index]/10;
                ans[index]=ans[index]%10;
            }
            if(carry>0){
                int index = num2.length()-1-i+num1.length();
                ans[index]+=carry;
                ans[index+1]+=ans[index]/10;
                ans[index]%=10;
            }
        }
        int index = num2.length()+num1.length();
        if(ans[index]>=10){
            ans[index+1]+=ans[index]/10;
            ans[index]%=10;
        }
        boolean start=false;
        String res="";
        for(int i=index+1;i>=0;i--){
            if(start || (ans[i]>0)){
                res=res+(char)(ans[i]+'0');
                start=true;
            }
        }
        if(res.equals("")) return "0";
        return res;
    }
}

File: Algorithms/43. Multiply Strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if len(num1)<len(num2):
            num1, num2 = num2, num1
        ans = [0 for i in range(len(num1)+len(num2)+1)]
        for i in range(len(num2)-1,-1,-1):
            temp = [0 for j in range(len(num1)+1)]
            cur = 0
            for j in range(len(num1)-1,-1,-1):
                temp[cur]=temp[cur]+(ord(num1[j])-ord('0'))*(ord(num2[i])-ord('0'))
                temp[cur+1]+=temp[cur]//10
                temp[cur]%=10
                cur+=1
            diff=len(num2)-1-i
            for j in range(len(num1)+1):
                ans[diff+j]+=temp[j]
                ans[diff+j+1]+=ans[diff+j]//10
                ans[diff+j]%=10
        while(ans and ans[-1]==0):
            ans.pop()
        if ans:
            res = ""
            ans.reverse()
            for x in ans:
                res=res+chr(ord('0')+x)
            return res
        else:
            return "0"